File: TopCoder/src/main/java/com/leetcode/tree/ConvertSortedListToBst.java
package com.leetcode.tree;

import com.leetcode.array.leetcode.RotateList;

//https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/
public class ConvertSortedListToBst {

    public TreeNode sortedListToBST(RotateList.ListNode head) {
        TreeNode node = helper(head);
        return node;
    }

    private TreeNode helper(RotateList.ListNode head) {
        if (head == null) {
            return null;
        }

        if (head.next == null) {
            return new TreeNode(head.val);
        }

        RotateList.ListNode midNode =  getMid(head);
        TreeNode root = new TreeNode(midNode.val);
        root.left = helper(head);
        root.right = helper(midNode.next);
        return root;
    }

    private RotateList.ListNode getMid(RotateList.ListNode head) {
        RotateList.ListNode slow = head;
        RotateList.ListNode fast = head;
        RotateList.ListNode pre = null;

        while (fast != null && fast.next != null) {
            pre = slow;
            slow = slow.next;
            fast = fast.next.next;
        }
        pre.next = null;
        return slow;
    }

    public static void main(String[] args) {
        RotateList.ListNode head = new RotateList.ListNode(-10);
        head.next = new RotateList.ListNode(-3);
        head.next.next = new RotateList.ListNode(0);
        head.next.next.next = new RotateList.ListNode(5);
        head.next.next.next.next = new RotateList.ListNode(9);
        TreeNode rootNode = new ConvertSortedListToBst().sortedListToBST(head);
        System.out.println(rootNode);
    }
}

File: TopCoder/python/ntree/ConvertSortedListToBst.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None



class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:

        result = self.helper(head)
        return result

    def helper(self, head: ListNode) -> TreeNode:
        if not head:
            return None

        if not head.next:
            return TreeNode(head.val)

        midNode = self.getMid(head)
        treeNode = TreeNode(midNode.val)
        treeNode.left = self.helper(head)
        treeNode.right = self.helper(midNode.next)

        return treeNode


    def getMid(self, head: ListNode) -> ListNode:
        fast = head
        slow = head
        pre = None

        while fast and fast.next:
            pre = slow
            slow = slow.next
            fast = fast.next.next

        pre.next = None
        return slow

if __name__ == '__main__':
    head = ListNode(-10)
    head.next = ListNode(-3)
    head.next.next = ListNode(0)
    head.next.next.next = ListNode(5)
    head.next.next.next.next = ListNode(9)

    sol = Solution()
    root = sol.sortedListToBST(head)
    print(root)
